src/payment/strategies/googlepay/googlepay-payment-strategy.ts

```diff
@@ -0,0 +1,371 @@
+            }).then( (response: GooglePayIsReadyToPayResponse) => {
+                if (response) {
+                    this._googlePaymentsClient.loadPaymentData(this._googlePaymentDataRequest)
+                        .then((paymentData: GooglePaymentData) => {
+                            return this._setExternalCheckoutData(paymentData);
+                        }).catch((err: GooglePaymentsError) => {
+                        reject(new Error(err.statusCode));
+                    });
+                }
+            });
+        });
+    }
+
+    private _createOrder(payment: Payment, useStoreCredit?: boolean, options?: PaymentRequestOptions): Promise<InternalCheckoutSelectors> {
+        return this._store.dispatch(this._orderActionCreator.submitOrder({ useStoreCredit }, options))
+            .then(() => this._store.dispatch(this._paymentActionCreator.submitPayment(payment)));
+    }
+
+    private _getGooglePaymentsClient(google: GooglePaySDK, testMode: boolean | undefined): GooglePayClient {
+        let environment: EnvironmentType;
+        testMode = true; // TODO: remove when push this code to final review
+        if (testMode === undefined) {
+            throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
+        }
+
+        if (!testMode) {
+            environment = 'PRODUCTION';
+        } else {
+            environment = 'TEST';
+        }
+
+        const options: GooglePayPaymentOptions = { environment };
+
+        return new google.payments.api.PaymentsClient(options);
+    }
+
+    private _setExternalCheckoutData(paymentData: GooglePaymentData): Promise<void> {
+        return this._googlePayInitializer.parseResponse(paymentData)
+            .then((tokenizePayload: TokenizePayload) => {
+                const paymentSuccessPayload: PaymentSuccessPayload = {
+                    tokenizePayload,
+                    billingAddress: paymentData.cardInfo.billingAddress,
+                    shippingAddress: paymentData.shippingAddress,
+                    email: paymentData.email,
+                };
+
+                const {
+                    onError = () => {},
+                    onPaymentSelect = () => {},
+                } = this._googlePayOptions;
+
+                this._updateShippingAndBillingAddress(paymentSuccessPayload).then(() => {
+                    return this._paymentInstrumentSelected(paymentSuccessPayload)
+                        .then(() => onPaymentSelect())
+                        .catch(error => onError(error));
+                });
+            });
+    }
+
+    private _paymentInstrumentSelected(paymentSuccessPayload: PaymentSuccessPayload): Promise<InternalCheckoutSelectors> {
+        if (!this._paymentMethod) {
+            throw new Error('Payment method not initialized');
+        }
+
+        const { id: methodId } = this._paymentMethod;
+
+        return this._store.dispatch(this._paymentStrategyActionCreator.widgetInteraction(() => {
```

Comment:
You need to wrap the `widgetInteraction` around all asynchronous calls, so including `_updateShippingAndBillingAddress`. At the moment, it sits outside. Therefore, it will emit the wrong pending status.

Reply:
@clopezh Can you acknowledge this comment and show the change you made to put it in place.